### ISSUE TITLE
PostgreSQL partitioning breaks subset of create statements and after_save callback

### ISSUE THREAD

**Issue by gdeglin** [rails/rails#21659]
PostgreSQL has built-in functionality to partition tables. Partitioning is an good way to maintain fast query speeds as tables increase in size: http://www.postgresql.org/docs/9.4/static/ddl-partitioning.html 

After enabling partitioning on a table, we noticed that the following Rails features broke:
1) Calling model_instance.create_[associated_record_name] deletes the new record immediately after creating it. 

Looking into this, the problem is related to `target` incorrectly not being equal to `record` on the following line when partitioning is enabled: https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/has_one_association.rb#L37

This appears to occur because of the Postgres trigger for partitioning returns NULL instead of NEW, which slightly changes the output of the insert statement. However, if the trigger is modified to return NEW, then the row will be duplicated across both the master table and the partition. This may be related to this issue: https://github.com/rails/rails/issues/1080

We were able to work around this problem by returning NEW from the trigger instead of NULL. We then creating a second trigger that deletes the duplicate row in the same transaction.

2) after_save is no longer called, but after_create is. The workaround for the problem was the same as above.

These were the only two issues we noticed across our application, so hopefully there isn't too much to fix here.

**Comment by blunney1** [comment]
And the database will suffer table and index bloat because of this approach, reducing performance ,

**Comment by gdeglin** [comment]
Here is a failing test case: https://gist.github.com/gdeglin/7be2e39ea29cb59c1409

In the output you can see the following

From the following lines
```
puts "------Before create profile"
user.create_profile
puts "------After create profile"
```

Result:
```
------Before create profile
D, [2015-09-20T16:23:48.851068 #62282] DEBUG -- :   SQL (0.6ms)  INSERT INTO "profiles" ("user_id") VALUES ($1) RETURNING "id"  [["user_id", 1]]
D, [2015-09-20T16:23:48.853510 #62282] DEBUG -- :   Profile Load (0.5ms)  SELECT  "profiles".* FROM "profiles" WHERE "profiles"."user_id" = $1 LIMIT 1  [["user_id", 1]]
D, [2015-09-20T16:23:48.855612 #62282] DEBUG -- :   SQL (0.4ms)  UPDATE "profiles" SET "user_id" = $1 WHERE "profiles"."id" = $2  [["user_id", nil], ["id", 1]]
------After create profile
```

So it looks like profiles is being inserted correctly (user_id = 1), but then immediately updated to an invalid value (user_id = null).

**Comment by gdeglin** [comment]
After some research I found that Rails already has a workaround for this problem, which is to use ```insert_returning: false``` in the database configuration, which switches to using sequences to determine the current id instead of trying to rely on 'returning id".

However I found two bugs in this workaround:
1) The workaround fails when calling create() with no parameters due to an invalid regex here https://github.com/rails/rails/blob/d0f0ec9b7d8beb93085af4772e46d7ef60a73d47/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb#L763. The regex fails to match ```INSERT INTO "users" DEFAULT VALUES```
2) The workaround fails for non autoincrementing primary keys like UUIDs.

Standby for a patch for the first bug. Still investigating the second one.
